domain/relation/state/relation.go

```diff
@@ -162,6 +162,85 @@ func (st *State) AddRelation(ctx context.Context, epIdentifier1, epIdentifier2 r
 	})
 }
 
+// ApplicationRelationsInfo returns all EndpointRelationData for an application.
+// If the application is not in any relations, no error is returned.
+//
+// The following error types can be expected to be returned:
+//   - [relationerrors.ApplicationNotFound] is returned if application does not
+//     exist.
+func (st *State) ApplicationRelationsInfo(
+	ctx context.Context,
+	appID application.ID,
+) ([]relation.EndpointRelationData, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var results []relation.EndpointRelationData
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		found, err := st.checkExistsByUUID(ctx, tx, "application", appID.String())
+		if err != nil {
+			return errors.Capture(err)
+		} else if !found {
+			return relationerrors.ApplicationNotFound
+		}
+
+		// Find every relation the application is part of.
+		relationIDUUIDAppNames, err := st.getEveryRelationForApplicationID(ctx, tx, appID)
+		if errors.Is(relationerrors.RelationNotFound, err) {
+			// The application is not in any relations.
+			return nil
+		} else if err != nil {
+			return errors.Capture(err)
+		}
+
+		// For each relation, get its EndpointsRelationData based on the application.
+		results = make([]relation.EndpointRelationData, len(relationIDUUIDAppNames))
+		for i, rel := range relationIDUUIDAppNames {
```

Comment:
One potato... 

We should be doing this in one query, and normalising into a data structure. That would hit the database a deterministic number of times rather than O(n).